src/stcal/alignment/util.py

```diff
@@ -302,19 +303,49 @@ def _calculate_new_wcs(wcs: gwcs.wcs.WCS,
         transform=transform,
         input_frame=wcs.input_frame,
     )
-    axis_min_values, output_bounding_box = _get_axis_min_and_bounding_box(footprints, wcs_new)
+    axis_min_values, bbox = _get_axis_min_and_bounding_box(footprints, wcs_new)
     offsets = _calculate_offsets(
         fiducial=fiducial,
         wcs=wcs_new,
         axis_min_values=axis_min_values,
         crpix=crpix,
     )
 
+    if crpix is None:
+        output_bounding_box = bbox
+    else:
+        output_bounding_box = []
+        for axis_range, minval, shift in zip(bbox, axis_min_values, crpix):
+            output_bounding_box.append(
+                (
+                    axis_range[0] + shift + minval,
+                    axis_range[1] + shift + minval
+                )
+            )
+
     wcs_new.insert_transform("detector", offsets, after=True)
     wcs_new.bounding_box = output_bounding_box
 
     if shape is None:
-        shape = [int(axs[1] - axs[0] + 0.5) for axs in output_bounding_box[::-1]]
+        if crpix is None:
+            shape = [
+                int(axs[1] - axs[0] + 0.5) for axs in output_bounding_box[::-1]
+            ]
+        else:
+            shape = []
+            for k, axs in enumerate(output_bounding_box[::-1]):
+                upper = int(axs[1] + 0.5)
+                if upper < 1:
+                    log.warning(
+                        "Input images do not overlap with created WCS. "
+                        "Consider adjusting crval and/or crpix values."
+                    )
+                    log.warning(
+                        "Setting minimum array dimension for axis %d to 10."
+                        % (len(output_bounding_box) - k)
+                    )
+                    upper = 10
```

Comment:
why is 10 the default here?

In what cases would upper be less than 1?  Should this ever happen for valid values of crval and crpix?  I thought this PR attempted to fix this lack of overlap for any reasonable input WCS.

Reply:
This PR initially attempted to fix lack of overlap _due to a bug_ in the code: applying `crpix` to coordinate transforms in WCS but not to the bounding box. However, some tests in `romancal` failed when the following two conditions occurred:
1. the combination of `crval` is such that all input images create an outline that is way outside of the bounding box (towards the negative coordinates); AND
2. shape is not specified.

When `crpix` is provided (but `size` is not), it could happen that only a small part (or none) is in the quadrant with all positive coordinates. Also, the bounding box could be shifted to the right and up (from the origin of the image coordinate system). In this case the size should be determined from the upper bounding box coordinates and not from the bounding box width and height. However, if the bounding box is, for example, in the quadrant with all negative coordinates, then what should be the size of the output image. I think `drizzle` needs a minimum of 2 or 3 pixels. 10 is a small enough number that it will not use a lot of resources but we can lower it to let's say 3.

Or would you prefer to raise an exception and quit?

Reply:
Hm, I think that is how `size` should be determined at all times (from the largest coordinate values of the bounding box): it just happens that the lower limit of the bounding box is always 0 when `crval` is not specified.

Reply:
Ok here is my understanding so far.  I still don't understand WCS transforms very well so please pardon me if anything here is obviously wrong:

- This PR is making it so that the `bounding_box` is updated to respect `crpix != 0`, which means that the forward and reverse transforms should be self-consistent and will respect `bounding_box`.
- there's no reason why `crpix` and `crval` cannot be outside the image itself, but `bounding_box` should always be around the image
- if `bounding_box` is nowhere near the image, then there is something wrong with either `bounding_box` or `crpix` and `crval`

Are these three statements true?
If so, isn't (1) an instance of a bad WCS, since it's making a `bounding_box` that isn't in the image?

Reply:
> * This PR is making it so that the `bounding_box` is updated to respect `crpix != 0`, which means that the forward and reverse transforms should be self-consistent and will respect `bounding_box`.

_"forward and reverse transforms should be self-consistent"_ - that is something that _always_ should be happening. _"will respect `bounding_box`"_ this depends on the settings such as `with_bounding_box` (for the inverse) and I think always for the forward transform. This PR does not deal with this.

> * there's no reason why `crpix` and `crval` cannot be outside the image itself, but `bounding_box` should always be around the image

First part of the statement is correct. Second part is correct too _if_ one wants to get finite numbers (not `nan`) for the world coordinates _for pixels in an actual image described by said WCS object_ or if one wants to get any data in the resampled image. If the bounding box is not in the image then pixel coordinates of that image's pixels will not map to finite world coordinates.

> * if `bounding_box` is nowhere near the image, then there is something wrong with either `bounding_box` or `crpix` and `crval`

"something wrong" - it depends. If this is the result of our computations for default parameters - then yes. If it is the result of the user's input - then it's user's choice (maybe by mistake or maybe not ... who knows?

> Are these three statements true? If so, isn't (1) an instance of a bad WCS, since it's making a `bounding_box` that isn't in the image?

What is (1) referring to? A WCS cannot make a `bounding_box`. We create a bounding box to enclose the data. That is what the current code does _before_ adding a shift to the detector coordinates to account for `crpix`.

At least for the case of resampled image, the bounding box should enclose the region of the image/detector coordinate system where all input images will be resampled (projected) but it also should be trimmed to the actual size of the images (a new thing to add to the code).

Reply:
> "forward and reverse transforms should be self-consistent" - that is something that always should be happening. "will respect bounding_box" this depends on the settings such as with_bounding_box (for the inverse) and I think always for the forward transform. This PR does not deal with this.

On and around [this line](https://github.com/spacetelescope/jwst/blob/2968d3f144f945a483f8c92fc67e4c52d11a91df/jwst/resample/resample_utils.py#L172) it appears that currently the `forward_transform` and `backward_transform` do _not_ round-trip if the `bounding_box` is not None.  I thought this PR was supposed to fix that.

> if one wants to get finite numbers (not nan) for the world coordinates for pixels in an actual image described by said WCS object or if one wants to get any data in the resampled image. If the bounding box is not in the image then pixel coordinates of that image's pixels will not map to finite world coordinates.

Can you give me an example of when getting NaN for the world coordinates on purpose, or getting no data in the resampled image, would be the desired behavior?

> If it is the result of the user's input - then it's user's choice (maybe by mistake or maybe not ... who knows?

Along the same lines, under what circumstances would this not be a mistake?

> What is (1) referring to? A WCS cannot make a bounding_box. We create a bounding box to enclose the data. That is what the current code does before adding a shift to the detector coordinates to account for crpix.

It was referring to your previous message, specifically:

> the combination of crval is such that all input images create an outline (bounding box) that is way outside of the output image (towards the negative coordinates);

I know that the WCS doesn't make the bounding box.  I meant, "the WCS is defined such that the bounding box has no overlap with the image"

Reply:
> On and around [this line](https://github.com/spacetelescope/jwst/blob/2968d3f144f945a483f8c92fc67e4c52d11a91df/jwst/resample/resample_utils.py#L172) it appears that currently the `forward_transform` and `backward_transform` do _not_ round-trip if the `bounding_box` is not None. I thought this PR was supposed to fix that.

You say "currently the `forward_transform` and `backward_transform` do _not_ round-trip". To me "round-tripping is that the following is true:

```python
world = forward_transform(x, y)
assert np.allclose(backward_transform(*world), (x, y))
```

Is this condition not satisfied (modulo `nan`)? I think it still should be. The issue is that the bounding box is off when `crpix` is specified by the user and hence `nan`s in the output. I believe this PR fixes the issue and so those lines in the `reproject` will not be needed.

> Can you give me an example of when getting NaN for the world coordinates on purpose, or getting no data in the resampled image, would be the desired behavior?

No. And that is the issue! Current code creates a bounding box (when `crpix` is defined) that may not be in the resampled => no data.

> Along the same lines, under what circumstances would this not be a mistake?

I don't know.

Reply:
ok, thanks for all the explanations, I think I understand now.  The bounding box does not modify the transform in any way, but it does set the results of the transform to NaN outside the box.

Going all the way back to the original question, if we can't think of any real use-cases for a bounding box that is entirely outside the data, then IMO it would be better to raise an exception than log a warning.

I will admit also that some of my confusion came because I made a typo when removing those lines in `resample_utils.reproject()` and it led to a bunch of test failures.  The branch I made (typo fixed) is [here](https://github.com/spacetelescope/jwst/compare/main...emolter:jwst:mihai-bbox), and I'm running our full regtest suite on those changes [here](https://github.com/spacetelescope/RegressionTests/actions/runs/12434172392), pointing to this PR branch as the stcal dependency.  I'm ready to approve this as long as those tests come back clean

Reply:
If you did change this `log.warning()` to an error, would that cause the Romancal tests to fail again?

Reply:
Can you help me understand why there are regression test failures in the run I linked above?

Reply:
> Can you help me understand why there are regression test failures in the run I linked above?

probably my prs were not rebased. 

I removed log warnings and now raise an exception.